pkg/virtctl/adm/logverbosity/logverbosity.go

```diff
@@ -0,0 +1,304 @@
+)
+
+const (
+	LOG_VERBOSITY = "log-verbosity"
+)
+
+type Command struct {
+	clientConfig clientcmd.ClientConfig
+	command      string
+}
+
+var (
+	virtAPIFlag        bool
+	virtControllerFlag bool
+	virtHanderFlag     bool
+	virtLauncherFlag   bool
+	virtOperatorFlag   bool
+	allFlag            bool
+	resetFlag          bool
+)
+
+func NewCommand(clientConfig clientcmd.ClientConfig) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:     "log-verbosity",
+		Short:   "Set/Reset log verbosity level",
+		Long:    "Set log verbosity level to one or more compoents (virt-api/virt-controller/virt-handler/virt-llauncher/virt-operator). Reset the log verbosity level to all components.",
+		Example: usage(),
+		Args:    cobra.MaximumNArgs(1),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			c := Command{command: LOG_VERBOSITY, clientConfig: clientConfig}
+			return c.RunE(args)
+		},
+	}
+
+	cmd.Flags().BoolVar(&virtAPIFlag, "virt-api", virtAPIFlag, "Specify virt-api log verbosity level.")
+	cmd.Flags().BoolVar(&virtControllerFlag, "virt-controller", virtControllerFlag, "Specify virt-controller log verbosity level.")
+	cmd.Flags().BoolVar(&virtHanderFlag, "virt-handler", virtHanderFlag, "Specify virt-handler log verbosity level.")
+	cmd.Flags().BoolVar(&virtLauncherFlag, "virt-launcher", virtLauncherFlag, "Specify virt-launcher log verbosity level.")
+	cmd.Flags().BoolVar(&virtOperatorFlag, "virt-operator", virtOperatorFlag, "Specify virt-operator log verbosity level.")
+	cmd.Flags().BoolVar(&allFlag, "all", allFlag, "Set all component log verbosity levels.")
+	cmd.Flags().BoolVar(&resetFlag, "reset", resetFlag, "Reset log verbosity level.")
+
+	// no required flag
+
+	cmd.SetUsageTemplate(templates.UsageTemplate())
+	return cmd
+}
+
+func usage() string {
+	usage := "  # Set a log verbosity level to a component:\n"
+	usage += "  {{ProgramName}} adm logVerbosity 3 --virt-api"
+	usage += "  # Set a log verbosity level to multiple components:\n"
+	usage += "  {{ProgramName}} adm logVerbosity 3 --virt-api  --virt-handler"
+	usage += "  # Set a log verbosity level to all components:\n"
+	usage += "  {{ProgramName}} adm logVerbosity 3 --all"
+	usage += "  # Reset the log verbosity level to all components:\n"
+	usage += "  {{ProgramName}} adm logVerbosity --reset"
+
+	return usage
+}
+
+func (c *Command) RunE(args []string) error {
+	/*
+	 * flag strength: reset > all > component
+	 */
+
+	var logLevel uint
+
+	// parse argument
+	if !resetFlag {
+		if len(args) == 0 {
+			return fmt.Errorf("expecting log verbosity level")
+		}
+		level, err := strconv.ParseUint(args[0], 10, 0)
+		if err != nil {
+			return err
+		}
+		logLevel = uint(level)
+
+		/*
+		 * user can specify logLevel 0
+		 * if we do not allow user to specify logLevel 0, need error handling here.
+		 */
+
+	}
+
+	// adm command always sees the KubeVirt resource of namespace="kubevirt", name="kubevirt"
+	const (
+		NAMESPACE = "kubevirt"
+		NAME      = "kubevirt"
+	)
```

Comment:
Please, define the costants at the beginning of the file

Reply:
I don't think you can always assume that the namespace and the name are `kubevirt`. For example, this can be different in Openshift installation

Reply:
Good catch. I was wondering if I could always assume that the namespace and name are "kubevirt". So I put these constants here temporarily.

I want to focus on the KubeVirt CR first, with the possibility of future expansion. So let me move the constants to the beginning for now.

I have an idea to get namespace and name like this.
```
namespace, _, err := clientConfig.Namespace()
kvList, err := virtClient.KubeVirt(namespace).List(&k8smetav1.ListOptions{})
kv := kvList.Items[0]
name = kv.Name
```

When I "make cluster-up" on my local environment, I failed to get the namespace and name by using the above method. I'll fix it and get the namespace and name, and remove the constants later.

Reply:
What you could do is listing KubeVirt CRs in all namespaces and see where it is installed (more or less the same thing as `kubectl get kubevirt -A`)

Reply:
Many thanks. This helps a lot. I can do something like this.

```
func detectInstallNamespaceAndName() (string, string) {
        kvs, err := virtClient.KubeVirt("").List(&metav1.ListOptions{})
        util.PanicOnError(err)
        if len(kvs.Items) == 0 {
                util.PanicOnError(fmt.Errorf("Could not detect a kubevirt installation"))
        }
        if len(kvs.Items) > 1 {
                util.PanicOnError(fmt.Errorf("Invalid kubevirt installation, more than one KubeVirt resource found"))
        }
        return kvs.Items[0].Namespace, kvs.Items[0].Name
}

namespace, name, err := detectInstallNamespaceAndName()
```

Reply:
Yes, if you search in kubevirt code, there should be something similar where we only allow a single instance of KubeVirt CR

Reply:
Yes, I was referring to this code when I wrote the snippet above.
https://sourcegraph.com/github.com/kubevirt/kubevirt@main/-/blob/tests/testsuite/namespace.go?L339-350

I need to understand more about the concept (the whole KubeVirt design) behind the code.